src/prototype_room_creepbuilder.js

```diff
@@ -67,34 +79,29 @@ Room.prototype.inQueue = function(creepMemory) {
  * @return {boolean}           if the spawn is not allow, it will return false.
  */
 Room.prototype.checkRoleToSpawn = function(role, amount, targetId, targetRoom, level, base) {
-  targetRoom = targetRoom || this.name;
-  amount = amount || 1;
-
-  let creepMemory = {
-    role: role,
-    level: level,
-    base: base || undefined,
-    routing: {
-      targetRoom: targetRoom,
-      targetId: targetId
-    }
-  };
+  var creepMemory = this.creepMem(role, targetId, targetRoom, level, base);
+  var creeps = this.find(FIND_MY_CREEPS);
+  var spawns = this.find(FIND_MY_SPAWNS);
   if (this.inQueue(creepMemory)) {return false;}
-  let creeps = this.find(FIND_MY_CREEPS);
-  let spawns = this.find(FIND_MY_SPAWNS);
-  for (let spawn of spawns) {
+  let spawn = {};
+  for (spawn of spawns) {
     if (!spawn.spawning) {continue;}
     creeps.push(Game.creeps[spawn.spawning.name]);
   }
-  creeps = _.filter(creeps, creep => {
-    if (!creep.memory.routing) {return false;}
-    let creepTarget = {targetId: creep.memory.routing.targetId,
-      targetRoom: creep.memory.routing.targetRoom};
-    return _.eq(creepMemory.routing, creepTarget) && role === creep.memory.role;
-  });
-  if (creeps.length < amount) {
-    this.memory.queue.push(creepMemory);
+  let j = 0;
+  let iMax = creeps.length || 0;
+  amount = amount || 1;
+  for (let i = 0; iMax; i++) {
+    let iMem = creeps[i].memory;
+    if (creepMemory.routing.targetRoom === iMem.routing.targetRoom &&
+       creepMemory.routing.targetId === iMem.routing.targetId &&
+       role === iMem.role) {
+      j++;
+    }
+    if (j >= amount) {return false;}
```

Comment:
the idea here was to reduce the cost for check spawn. i also modify it more for register when there is enouth in queue for wait some ticks before check again. This PR is now unstructured so i'll surely close it and do differents modifications in differents PR.